### PR TITLE
[TVMC] add cl support in tvmc runner

### PR DESCRIPTION
**Pull request by euntaik**
Thanks for contributing to TVM!   Please refer to guideline https://tvm.apache.org/docs/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/apache/incubator-tvm/blob/master/CONTRIBUTORS.md#reviewers) by @ them in the pull request thread.

### CODE REVIEW

**Review by leandron — CHANGES_REQUESTED** [review]
Hi @euntaik, thanks for the patch. Adding support to CL is something we are looking forward.

Could you also add some tests, to make sure that support doesn't break in the future without us noticing? There are two kinds tests I think we need here: some for compilation targeting `cl`, as well as runner tests.

**Comment by leandron** [comment]
cc @comaniac

**Review by comaniac — CHANGES_REQUESTED** [review]
Please address the comments from @leandron

File: python/tvm/driver/tvmc/runner.py [hunk 1]
```diff
@@ -50,7 +50,7 @@ def add_run_parser(subparsers):
     #      like 'cl', 'webgpu', etc (@leandron)
```
Comment (leandron) [inline python/tvm/driver/tvmc/runner.py:50]:
```suggestion
    #      like 'webgpu', etc (@leandron)
```

File: python/tvm/driver/tvmc/compiler.py [hunk 2]
```diff
@@ -185,11 +185,11 @@ def compile_model(
         with autotvm.apply_history_best(tuning_records):
             with tvm.transform.PassContext(opt_level=3):
                 logger.debug("building relay graph with tuning records")
-                graph_module = relay.build(mod, tvm_target, params=params, target_host=tvm_target)
+                graph_module = relay.build(mod, tvm_target, params=params, target_host=target_host)
```
Comment (leandron) [inline python/tvm/driver/tvmc/compiler.py:188]: This looks unrelated to the overall CL addition, I guess as it is a minor variable replacement fix, we can leave it here - I'm just pointing it out.

File: tests/python/driver/tvmc/test_runner.py [hunk 3]
```diff
@@ -96,3 +96,14 @@ def test_run_tflite_module__with_profile__valid_input(
...
+
+def test_run_tflite_module_on_opencl(tflite_opencl_compiled_module_as_tarfile):
+    # some CI environments wont offer TFLite, so skip in case it is not present
+    pytest.importorskip("tflite")
+
+    outputs, times = tvmc.runner.run_module(
+        tflite_opencl_compiled_module_as_tarfile, hostname=None, device="cl"
+    )
+    assert type(outputs) is dict
+    assert type(times) is tuple
```
Comment (leandron) [inline tests/python/driver/tvmc/test_runner.py:109]: Can you add one more check, to see whether the output tensor was produced? (see line 98 for an example)

**Review by leandron — CHANGES_REQUESTED** [review]

File: tests/python/driver/tvmc/conftest.py [hunk 4]
```diff
@@ -70,6 +89,19 @@ def tflite_mobilenet_v1_1_quant(tmpdir_factory):
     return model_file
 
 
+@pytest.fixture(scope="session")
+def tflite_mobilenet_v1_0_25_128(tmpdir_factory):
```
[the remark this answers is not in the document]
Comment (euntaik) [reply tests/python/driver/tvmc/conftest.py:92]: I wanted to use the smallest possible mobilenet without the quantization.

File: tests/python/driver/tvmc/conftest.py [hunk 5]
```diff
@@ -54,6 +54,25 @@ def get_sample_compiled_module(target_dir):
...
+        "{}/{}".format(base_url, model_url),
+        "mobilenet_v1_0.25_128.tflite",
+        temp_dir=target_dir,
+    )
+    return tvmc.compiler.compile_model(
+        model_file,
+        target="opencl",
+        target_host="llvm",
+        model_format="tflite",
+        dump_code="ll",
```
[the remark this answers is not in the document]
Comment (euntaik) [reply tests/python/driver/tvmc/conftest.py]: No I don't need it. Committed the change.

File: tests/python/driver/tvmc/test_runner.py (hunk 3 above)
Comment (euntaik) [reply tests/python/driver/tvmc/test_runner.py:110]: Added the output check.

File: tests/scripts/task_config_build_cpu.sh [hunk 6]
```diff
@@ -45,3 +45,4 @@ echo set\(USE_TENSORFLOW_PATH \"/tensorflow\"\) >> config.cmake
 echo set\(USE_FLATBUFFERS_PATH \"/flatbuffers\"\) >> config.cmake
 echo set\(USE_ETHOSN /opt/arm/ethosn-driver\) >> config.cmake
 echo set\(USE_ETHOSN_HW OFF\) >> config.cmake
+echo set\(USE_OPENCL ON\) >> config.cmake
```
Comment (leandron) [inline tests/scripts/task_config_build_cpu.sh:48]: This will need to go as a separate PR, and be merged before you submit this one.

Please create a new PR and cc @tqchen

**Review by leandron — CHANGES_REQUESTED** [review]

File: tests/scripts/task_config_build_cpu.sh (hunk 6 above)
Comment (euntaik) [reply tests/scripts/task_config_build_cpu.sh:48]: Created a PR https://github.com/apache/incubator-tvm/pull/6847/files

File: python/tvm/driver/tvmc/compiler.py (hunk 2 above)
Comment (comaniac) [reply python/tvm/driver/tvmc/compiler.py:189]: @leandron please be aware that we should include this fix in the back port, but we will not include this PR since this is a new feature. Accordingly, I'd suggest separating this fix.

File: python/tvm/driver/tvmc/runner.py [hunk 7]
```diff
@@ -361,7 +361,12 @@ def run_module(
 
         # TODO expand to other supported devices, as listed in tvm.rpc.client (@leandron)
         logger.debug("device is %s", device)
-        ctx = session.cpu() if device == "cpu" else session.gpu()
+        if device == "gpu":
+            ctx = session.gpu()
+        elif device == "cl":
+            ctx = session.cl()
+        else:
+            ctx = session.cpu()
```
Comment (comaniac) [inline python/tvm/driver/tvmc/runner.py:369]:
```suggestion
            assert device == "cpu"
            ctx = session.cpu()
```

File: tests/python/driver/tvmc/conftest.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (comaniac) [reply tests/python/driver/tvmc/conftest.py:92]: Same opinion. I don't see a strong reason of adding an end-to-end model test. Please be aware that running e2e network in the unit test is relatively time-consuming, so we should avoid unnecessary test as possible.

**Review by comaniac — CHANGES_REQUESTED** [review]
IMHO, unit tests should be isolated and only test the target components. Specifically, the unit tests for TVMC should not test the entire compilation and execution flow but just need to make sure the inputs to other TVM components are correct. Whether the codegen or runtime are successes or not should not be covered in TVMC unit tests (that means, this PR should not enable OpenCL in the CI).

@leandron I realized this issue also applies to some existing tests. Could we review the TVMC unit tests again later and make them more precise and concise? If we found that the reason to have such tests here is that this is never covered else where, it is totally fine to add more unit tests to the right place (e.g., codegen, runtime).

**Comment by euntaik** [comment]
created a new PR for changes in compiler.py
https://github.com/apache/incubator-tvm/pull/6855

**Comment by leandron** [comment]
> @leandron I realized this issue also applies to some existing tests. Could we review the TVMC unit tests again later and make them more precise and concise? If we found that the reason to have such tests here is that this is never covered else where, it is totally fine to add more unit tests to the right place (e.g., codegen, runtime).

I agree we don't need to run *all* tests as end-to-end, and I can remove some overlap, by creating some mocks and checking that calls are made to the internal APIs with the right parameters (as you mentioned), but I'd like a few to stay - I'll see to get it done, as soon as I can finish a PR I'm working on.

File: tests/python/driver/tvmc/conftest.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (leandron) [reply tests/python/driver/tvmc/conftest.py]: Yeah, this could be the first one to check only API calls, using a mock, rather than downloading a file and exercising the whole e2e path.

**Comment by comaniac** [comment]
> I agree we don't need to run _all_ tests as end-to-end, and I can remove some overlap, by creating some mocks and checking that calls are made to the internal APIs with the right parameters (as you mentioned), but I'd like a few to stay - I'll see to get it done, as soon as I can finish a PR I'm working on.

It makes sense to keep a few for reasons and we can discuss on the PR. Thanks :)

**Review by comaniac — APPROVED** [review]

**Comment by comaniac** [comment]
Thanks @euntaik @leandron. We can work on adding more tests along with other TVMC TODOs.
